File: dp_labs_and_solutions/dp_labs_and_solutions/dp_matlab_labs_and_solutions/template_method/solution/Purchase.m
% Purchase.m

classdef (Abstract) Purchase < handle
	methods (Sealed)
		function perform(thePurchase)
			thePurchase.selectItem;
			thePurchase.bringItemToRegister;
			thePurchase.pay;
		end
	end

	methods (Abstract, Access = protected)
		pay(thePurchase)
	end

	methods (Access = private, Sealed)
		function selectItem(thePurchase)
		end

		function bringItemToRegister(thePurchase)
		end
	end
end

File: dp_labs_and_solutions/original/dp_labs_and_solutions/dp_java_labs_and_solutions/template_method/solution/Purchase.java
// Purchase.java

public abstract class Purchase
{
	public final void perform()
	{
		selectItem();
		bringItemToRegister();
		pay();
	}

	private final void selectItem()
	{
	}

	private final void bringItemToRegister()
	{
	}

	protected abstract void pay();
}
